camel/agents/task_agent.py

```diff
@@ -181,3 +182,210 @@ def run(
+
+        system_message = BaseMessage(
+            role_name="Task Planner",
+            role_type=RoleType.ASSISTANT,
+            meta_dict=None,
+            content="You are a helpful task planner.",
+        )
+
+        super().__init__(system_message, model, model_config,
+                         output_language=output_language)
+
+        if (objective):
+            self.objective = objective
+            self.task_creator_prompt = self.task_creator_prompt.format(
+                objective=self.objective)
+
+    def run(
+        self,
+        previous_task: Union[str, TextPrompt],
+        task_result: Union[str, TextPrompt],
+        task_list: Optional[List[str]] = None,
+    ) -> List[dict]:
+        r"""Generate subtasks based on the previous task results and
+        incomplete task list.
+
+        Args:
+            previous_task (Union[str, TextPrompt]): The last completed task to
+                be used to create future plans.
+            task_result (Union[str, TextPrompt]): The result of last completed
+                task to be used to create future plans.
+            task_list (List[str]): The incompleted task list
+            which should not overlap with new created tasks.
+        Returns:
+            List[dict]: The new task list generated by the AI.
+        """
+        self.reset()
+        self.task_creator_prompt = self.task_creator_prompt.format(
+            task=previous_task, task_result=task_result)
+        if (task_list):
+            unsolved = f"These are unsolved tasks: {task_list}.\n\
+                These new tasks must not overlap with incomplete tasks."
+
+            self.task_creator_prompt = self.task_creator_prompt.format(
+                unsolved_tasks=unsolved)
+        else:
+            self.task_creator_prompt = self.task_creator_prompt.format(
+                unsolved_tasks="No incomplete tasks.")
+
+        task_msg = BaseMessage.make_user_message(
+            role_name="Task Planner", content=self.task_creator_prompt)
+
+        task_response = self.step(task_msg)
+
+        if task_response.msgs is None:
+            raise RuntimeError("Got None Subtasks messages.")
+        if task_response.terminated:
+            raise RuntimeError("Task planning failed.")
+
+        sub_tasks_msg = task_response.msgs[0]
+        new_tasks = sub_tasks_msg.content.strip().split('\n')
+        new_tasks_list = []
+        for task_string in new_tasks:
+            task_parts = task_string.strip().split(".", 1)
+            if len(task_parts) == 2:
+                # this may cause error as LLM may generated # instead of number
+                # task_id = ''.join(s for s in task_parts[0] if s.isnumeric())
+                task_name = re.sub(r'[^\w\s_]+', '', task_parts[1]).strip()
+                if task_name.strip():  # and task_id.isnumeric():
+                    new_tasks_list.append(task_name)
+        out = [{"task_name": task_name} for task_name in new_tasks_list]
```

Comment:
Just curious why we need return a list of dictionary instead of a list of strings?

Reply:
Same question. If the keys are the same, I guess a list of strings is sufficient.